theRAPTLab/gsgo
# Feature: Edit Instance Init Scripts using Wizard

Addresses #754

Instance init scripts in the Map Editor were originally edited directly with a code editor.  This adds the ability to edit instance init scripts in a Script Editor window, supporting both a wizard view and a code editor view.


# To Test

1. Open Main
2. Click on "SETUP" to edit the Map
3. Click on an instance in "CHARACTERS" (or if you haven't created an instance already, select blueprint from "ADD CHARACTERS")
4. Click on the instance to expand the Edit view.
5. Click on the "Edit" button to open the init script in a new wizard. 
6. Edit the script using a wizard as you normally would.
7. Click "Save Line" to save changes to the line.
8. Click "Save to Server" to update the init script.
9. The instance will be updated.

Other things to test:
* Delete the init script from Map Editor -- does anything break?
* Delete the init script from Script Editor -- does anything break?
* Drag and drop the instance in the Map Editor view -- are you able to update the position?  If the ScriptEditor window is open while you're dragging, does the code update?
* Make sure default initScript position code works with variations of `prop x setTo...` `prop agent.x setTo..` and `prop character.x setTo..`
* 


# Technical Background

Editing init script is a bit tricky because the wizard needs to pre-compile the blueprint in order to know how to scaffold the wizard editor. The blueprint script needs to be compiled and bundled so that the wizard knows which props and features have been defined and are therefore available to be called in the initScript.  We do this by adding initScripts to the blueprint bundle, and then re-purposing the ScriptEditor components to handle either regular blueprint scripts or just init scripts.


# To Do
* [X] ScriptEditor should indicate whether we're in "Character" edit mode or "Instance" edit mode?  How do we designate the difference since "instance" is not in the vocabulary?  Use "Character Type Editor" vs "Character Instance Editor"?  
* [X] Remove project name from ScriptEditor window
* [x] Use `<bpname>.x`for updating drag init scripts.
* [X] "Save to Server" is not clearing 'isDirty' flag
* [X] After Save Script to Server, character is not updated (alpha change is not reflected)
* [X] If ScriptEditor is open and the user drags, the code should update?  ~~Or do we lock out the drag?~~ We need to update otherwise the user will be left with an out of date ScriptEditor.
* [X] If initScript is deleted from MapEditor while ScriptEditor is open, close the window, don't revert to Blueprint edit view.
* [X] If initScript is deleted from MapEditor while ScriptEditor is open, line selection is broken.  Can we close the window?  Or at least properly revert, removing the initScript? => Window is now closed.
* [X] Changed bp name remains listed in "CHARACTER TYPE SCRIPTS"
* [X] "Children with the same key" error when editing alpha -- CAN'T REPLICATE or FIXED.  Won't do.
* [X] Bookmarks menu needs to be added if a bookmarkable comment is added. -- Fixed by 00a4022d9d4e5b4cc9aa65bd2a3828c1b2249aae
